pkg/operators/wasm/datasource.go

```diff
@@ -0,0 +1,297 @@
+	dsName, err := stringFromStack(m, stack[0])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	ds := i.gadgetCtx.GetDataSources()[dsName]
+	if ds == nil {
+		i.logger.Debugf("datasource not found %q", dsName)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(ds))
+}
+
+// dataSourceGetField returns a handle to a data source.
+// Params:
+// - stack[0]: Datasource handle
+// - stack[1]: Field name
+// Return value:
+// - Field handle
+func (i *wasmOperatorInstance) dataSourceGetField(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("datasource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, stack[1])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc := ds.GetField(fieldName)
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+// dataSourceAddField add a field to the data source and returns its handle.
+// Params:
+// - stack[0]: Datasource handle
+// - stack[1]: Field name
+// - stack[2]: Field kind
+// Return value:
+// - Field handle
+func (i *wasmOperatorInstance) dataSourceAddField(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("datasource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, stack[1])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc, err := ds.AddField(fieldName, api.Kind(stack[2]))
+	if err != nil {
+		i.logger.Warnf("adding field %q to datasource %q: %v", fieldName, ds.Name(), err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+// dataSourceSubscribe subscribes to the datasource.
+// Params:
+// - stack[0]: Datasource handle
+// - stack[1]: Priority
+// - stack[2]: Callback ID
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) dataSourceSubscribe(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("datasource handle %d not found", stack[0])
+		stack[0] = 1
+		return
+	}
+	prio := wapi.DecodeI32(stack[1])
+	cbID := stack[2]
+	cb := m.ExportedFunction("dsCallback")
+	if cb == nil {
+		i.logger.Warnf("module doesn't export dsCallback")
+		stack[0] = 1
+		return
+	}
+	ds.Subscribe(func(source datasource.DataSource, data datasource.Data) error {
+		// TODO: can it be done outside? Why does DataFunc need the datasource again?
+		tmpDS := i.addHandle(source)
+		tmpData := i.addHandle(data)
+		_, err := cb.Call(ctx, cbID, wapi.EncodeU32(tmpDS), wapi.EncodeU32(tmpData))
+		i.delHandle(tmpDS)
+		i.delHandle(tmpData)
+		return err
```

Comment:
Yes, we could use the outer source here to avoid adding/removing that handle all the time.

As to why: you could have the actual callback function as member of your `wasmOperatorInstance` struct and wouldn't need to store the context (the DataSource) somewhere, as the context is always given. It's more of a convenience thing and should be cheap enough (if not used exactly like it's used here, haha).